### PR TITLE
Windows installer and merge module implementation

### PR DESCRIPTION
**Pull request by samkearney**
Create a Wix installer and merge module for the RDMnet Broker service.

The merge module installs the Broker service, makes firewall exceptions for it, and installs a skeleton configuration file.

The installer is a thin MSI wrapper over the merge module which also brings in its dependencies (VCRedist and mDNSResponder)

### CODE REVIEW

**Comment by samkearney** [comment]
@bhair The stuff in `tools/install/windows` will be what's interesting to you. Thanks!

**Comment by bhair** [comment]
Is this not targetting x64? I see the config files are mentioning x86.

**Comment by samkearney** [comment]
@bhair 

> Is this not targetting x64? I see the config files are mentioning x86.

Correct. I've talked to @docsteer about this and we are going to distribute these services as 32-bit for maximum compatibility with old systems. This is considered a worthwhile design trade-off since it doesn't force any other apps to be 32-bit.

File: tools/install/windows/merge/BrokerMerge.wxs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    <Package Id="DEE166EC-5441-43DC-9532-FFAFB81FFA93" Description="$(var.MergeDescription)"
+        Manufacturer="ETC Inc." InstallerVersion="200" Languages="1033" SummaryCodepage="1252"
+        InstallScope="perMachine" InstallPrivileges="elevated" />
+
+    <!-- ############### Broker Service Directory Structure ############### -->
+
+    <!-- The executable directory in Program Files (x86) -->
+    <Directory Id="TARGETDIR" Name="SourceDir">
+      <Directory Id="$(var.ProgramFilesFolder)">
```
Comment (bhair) [inline tools/install/windows/merge/BrokerMerge.wxs:17]: Why are you using this var that's passed in from the cmd line when built? Wix has vars for both program files locations.
```
<Directory Id='TARGETDIR' Name='SourceDir'>
    <Directory Id='ProgramFiles64Folder'>
        ...
    </Directory>
    <Directory Id='ProgramFilesFolder'>
        ...
    </Directory>
</Directory>
```
I'm doing the following in Eos. You can have them go either place from a 32 or 64 bit WiX project.

File: tools/install/windows/merge/BrokerMerge.wxs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            <Component Id="Broker" Guid="2B4014E4-5278-469A-ABA7-A94F0AB850DA" SharedDllRefCount='yes'>
+              <File Id="BrokerExe" Name="RDMnetBrokerService.exe" DiskId="1" KeyPath="yes">
+                <!-- Create Windows Firewall exceptions for the Broker executable. -->
+                <fire:FirewallException Id="RDMnetBrokerTCP" Name="ETC RDMnet Broker (TCP-in)"
+                    Profile="all" Protocol="tcp" Scope="any" IgnoreFailure="no" />
+                <fire:FirewallException Id="RDMnetBrokerUDP" Name="ETC RDMnet Broker (UDP-in)"
+                    Profile="all" Protocol="udp" Scope="any" IgnoreFailure="no" />
+              </File>
+              <File Id="DnsSdDll" Name="dnssd.dll" DiskId="1" />
+              <ServiceControl Id="BrokerServiceCtl" Name="ETC RDMnet Broker" Stop="both"
```
Comment (bhair) [inline tools/install/windows/merge/BrokerMerge.wxs:29]: Haven't done a service yet, that seems almost too simple!

File: tools/install/windows/merge/BrokerMerge.wxs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    </Directory>
+
+    <!-- ######################### Custom Actions ######################### -->
+
+    <!--
+    The RDMnetBrokerService executable is invoked with the argument "-install"
+    on installation and "-remove" on uninstallation - this does the actual
+    registration and de-registration of the service with the Service Control
+    Manager.
+    -->
```
Comment (bhair) [inline tools/install/windows/merge/BrokerMerge.wxs:55]: Is this service registration unique to your service or did you find that this was the generally accepted way to go about this? I've not had to do services in an MSI yet so I'm curious.

File: tools/install/windows/standalone/BrokerStandalone.wxs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    required dependencies - the mDNSWindows and VCRedist merge modules.
+    -->
+    <Directory Id="TARGETDIR" Name="SourceDir">
+      <Merge Id="VcRedistx86" SourceFile="$(var.VcRedistMergeModulex86)" DiskId="1" Language="0" />
+      <Merge Id="mDNSResponder" SourceFile="$(var.mDNSMergeModule)" DiskId="1" Language="0" />
+      <Merge Id="EtcRdmnetBroker" SourceFile="..\merge\bin\$(var.Configuration)\EtcRdmnetBroker.msm"
+          DiskId="1" Language="0" />
+    </Directory>
+
+    <Feature Id="VcRedistx86Feature" Title="Visual C++ 32-bit Runtime" AllowAdvertise="no"
```
Comment (bhair) [inline tools/install/windows/standalone/BrokerStandalone.wxs]: You could have a single feature reference all of your merge modules. Don't have to have separate features for everything unless you want to give people the choice during install time of which features to install.

File: tools/install/windows/merge/BrokerMerge.wxs (hunk 1 above)
Comment (samkearney) [reply tools/install/windows/merge/BrokerMerge.wxs:17]: In vars.wxi we have:
```xml
  <?if $(var.Platform) = x64 ?>
    <?define ProgramFilesFolder = "ProgramFiles64Folder" ?>
    <?define MergeDescription = "ETC RDMnet Broker Service 64-bit Merge Module" ?>
    <?define ProductDescription = "ETC RDMnet Broker Service 64-bit Installer" ?>
  <?else ?>
    <?define ProgramFilesFolder = "ProgramFilesFolder" ?>
    <?define MergeDescription = "ETC RDMnet Broker Service 32-bit Merge Module" ?>
    <?define ProductDescription = "ETC RDMnet Broker Service 32-bit Installer" ?>
  <?endif ?>
```
So if I select the "x86" configuration in Visual Studio, it builds an MSI that installs the binaries in `ProgramFilesFolder`, but if I select the "x64" configuration it builds one that installs the binaries in `ProgramFiles64Folder`. Does that make sense?

File: tools/install/windows/merge/BrokerMerge.wxs (hunk 3 above)
Comment (samkearney) [reply tools/install/windows/merge/BrokerMerge.wxs:55]: This method (complete with the `-install` and `-remove` flags) seems to be the generally accepted practice in almost every service I've come across:
- The example service code available on MSDN
- Apple's Bonjour for Windows service
- Our SLP and SNTP service

**Comment by bhair** [comment]
Testing reply from email to see what happens.

Yes, that makes sense. I saw that file later on and figured you were doing something like that.

-Brian Hair

**Pull request merged**